src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+            if k.eq(word) {
+                handle = Some(*v);
+                break;
+            }
+        }
+        handle
+    }
+}
+
+#[derive(Debug, PartialEq)]
+struct LookupFunctionType {
+    parameter_type_ids: Vec<Word>,
+    return_type_id: Word,
+}
+
+pub struct Parser {
+    module: Module,
+    id_count: u32,
+    capabilities: Vec<Capability>,
+    debugs: Vec<Instruction>,
+    annotations: Vec<Instruction>,
+    debug_enabled: bool,
+    void_type: Option<u32>,
+    lookup_type: FastHashMap<Word, crate::Handle<crate::Type>>,
+    lookup_function: FastHashMap<Word, crate::Handle<crate::Function>>,
+    lookup_function_type: FastHashMap<Word, LookupFunctionType>,
+    lookup_constant: FastHashMap<Word, crate::Handle<crate::Constant>>,
+    lookup_global_variable: FastHashMap<Word, crate::Handle<crate::GlobalVariable>>,
+    lookup_import: FastHashMap<Word, String>,
+    lookup_label: Vec<Word>,
+    current_label: Word,
+}
+
+impl Parser {
+    pub fn new(module: &crate::Module, debug_enabled: bool) -> Self {
+        Parser {
+            module: Module::new(module),
+            id_count: 0,
+            capabilities: vec![],
+            debugs: vec![],
+            annotations: vec![],
+            debug_enabled,
+            void_type: None,
+            lookup_type: FastHashMap::default(),
+            lookup_function: FastHashMap::default(),
+            lookup_function_type: FastHashMap::default(),
+            lookup_constant: FastHashMap::default(),
+            lookup_global_variable: FastHashMap::default(),
+            lookup_import: FastHashMap::default(),
+            lookup_label: vec![],
+            current_label: 0,
+        }
+    }
+
+    fn generate_id(&mut self) -> Word {
+        self.id_count += 1;
+        self.id_count
+    }
+
+    fn bytes_to_words(&self, bytes: &[u8]) -> Vec<Word> {
+        let words: Vec<Word> = bytes
+            .chunks(4)
+            .map(|chars| match chars.len() {
+                4 => {
+                    (chars[3] as u32) << 24
+                        | (chars[2] as u32) << 16
+                        | (chars[1] as u32) << 8
+                        | chars[0] as u32
+                }
+                3 => {
+                    0x0u32 << 24
+                        | (chars[2] as u32) << 16
+                        | (chars[1] as u32) << 8
+                        | (chars[0] as u32) as u32
+                }
+                2 => 0x0u32 << 24 | 0x0u32 << 16 | (chars[1] as u32) << 8 | chars[0] as u32,
+                1 => 0x0u32 << 24 | 0x0u32 << 16 | 0x0u32 << 8 | chars[0] as u32,
+                _ => 0x0u32,
+            })
+            .collect();
+
+        words
+    }
+
+    fn string_to_words(&self, input: &str) -> Vec<Word> {
+        let mut words: Vec<Word> = self.bytes_to_words(input.as_bytes());
+
+        let last_word = words.last().unwrap();
+
+        let last_character = last_word.to_le_bytes()[3];
+
+        if last_character != 0x0 {
```

Comment:
alternatively, could check if `input.as_bytes() % 4 == 0`

Reply:
Not exactly sure, because that will only tell if the input is correctly aligned, or am I missing something?

Reply:
If the input is exactly 4 byte aligned, then last character is not 0.

Reply:
Thanks that is logical indeed. I made a small change, but I guess you meant this:
```
if input.as_bytes().len() %4 == 0
```